File: Backend/schemas/field.py
"""
field.py (schemas)

API-facing request/response schemas for Field and Crop Cycle endpoints.

Relationship chain:
    farmer → farm → fields → crop_cycles → crops

Validation rules are farmer-friendly:
  - Area must be >= 0
  - Lat/lng bounds are reasonable (not strict)
  - All fields except name are optional
"""

from pydantic import BaseModel, Field, field_validator
from typing import Optional


# ---------------------------------------------------------------------------
# Allowed values
# ---------------------------------------------------------------------------

ALLOWED_FIELD_STATUSES = {"active", "fallow", "inactive"}
ALLOWED_CYCLE_STATUSES = {"active", "harvested", "cancelled"}
ALLOWED_IRRIGATION_METHODS = {"canal", "tubewell", "drip", "sprinkler", "rainfed"}


# ---------------------------------------------------------------------------
# Field response / request schemas
# ---------------------------------------------------------------------------

class FieldResponse(BaseModel):
    """A single field returned to the frontend."""

    id: str
    farm_id: str
    name: str
    area_acres: Optional[float] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    boundary_geojson: Optional[str] = None
    soil_type: Optional[str] = None
    irrigation_method: Optional[str] = None
    status: str = "active"
    is_demo: bool = False

    # Nested active crop cycle (if any) for convenience.
    active_crop_cycle: Optional["CropCycleResponse"] = None


class FieldCreateRequest(BaseModel):
    """Payload for creating a new field."""

    name: str = Field(min_length=1, max_length=100)
    area_acres: Optional[float] = Field(default=None, ge=0, le=100000)
    latitude: Optional[float] = Field(default=None, ge=-90, le=90)
    longitude: Optional[float] = Field(default=None, ge=-180, le=180)
    boundary_geojson: Optional[str] = None
    soil_type: Optional[str] = Field(default=None, max_length=50)
    irrigation_method: Optional[str] = None
    status: str = "active"

    @field_validator("irrigation_method")
    @classmethod
    def validate_irrigation(cls, value: Optional[str]) -> Optional[str]:
        if value is not None and value not in ALLOWED_IRRIGATION_METHODS:
            raise ValueError(
                f"irrigation_method must be one of {sorted(ALLOWED_IRRIGATION_METHODS)}"
            )
        return value

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        if value not in ALLOWED_FIELD_STATUSES:
            raise ValueError(
                f"status must be one of {sorted(ALLOWED_FIELD_STATUSES)}"
            )
        return value


class FieldUpdateRequest(BaseModel):
    """Payload for updating an existing field (partial)."""

    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    area_acres: Optional[float] = Field(default=None, ge=0, le=100000)
    latitude: Optional[float] = Field(default=None, ge=-90, le=90)
    longitude: Optional[float] = Field(default=None, ge=-180, le=180)
    boundary_geojson: Optional[str] = None
    soil_type: Optional[str] = Field(default=None, max_length=50)
    irrigation_method: Optional[str] = None
    status: Optional[str] = None

    @field_validator("irrigation_method")
    @classmethod
    def validate_irrigation(cls, value: Optional[str]) -> Optional[str]:
        if value is not None and value not in ALLOWED_IRRIGATION_METHODS:
            raise ValueError(
                f"irrigation_method must be one of {sorted(ALLOWED_IRRIGATION_METHODS)}"
            )
        return value

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: Optional[str]) -> Optional[str]:
        if value is not None and value not in ALLOWED_FIELD_STATUSES:
            raise ValueError(
                f"status must be one of {sorted(ALLOWED_FIELD_STATUSES)}"
            )
        return value


# ---------------------------------------------------------------------------
# Crop cycle response / request schemas
# ---------------------------------------------------------------------------

class CropCycleResponse(BaseModel):
    """A single crop cycle returned to the frontend."""

    id: str
    field_id: str
    crop_name: str
    variety: Optional[str] = None
    crop_stage: Optional[str] = None
    planting_date: Optional[str] = None
    expected_harvest_date: Optional[str] = None
    status: str = "active"
    is_demo: bool = False


class CropCycleCreateRequest(BaseModel):
    """Payload for creating a new crop cycle on a field."""

    crop_name: str = Field(min_length=1, max_length=50)
    variety: Optional[str] = Field(default=None, max_length=50)
    crop_stage: Optional[str] = Field(default=None, max_length=50)
    planting_date: Optional[str] = None
    expected_harvest_date: Optional[str] = None
    status: str = "active"

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        if value not in ALLOWED_CYCLE_STATUSES:
            raise ValueError(
                f"status must be one of {sorted(ALLOWED_CYCLE_STATUSES)}"
            )
        return value


class CropCycleUpdateRequest(BaseModel):
    """Payload for updating an existing crop cycle (partial)."""

    crop_name: Optional[str] = Field(default=None, min_length=1, max_length=50)
    variety: Optional[str] = Field(default=None, max_length=50)
    crop_stage: Optional[str] = Field(default=None, max_length=50)
    planting_date: Optional[str] = None
    expected_harvest_date: Optional[str] = None
    status: Optional[str] = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: Optional[str]) -> Optional[str]:
        if value is not None and value not in ALLOWED_CYCLE_STATUSES:
            raise ValueError(
                f"status must be one of {sorted(ALLOWED_CYCLE_STATUSES)}"
            )
        return value


# ---------------------------------------------------------------------------
# Farm summary with fields (used by dashboard + map page)
# ---------------------------------------------------------------------------

class FarmWithFieldsResponse(BaseModel):
    """Farm overview including all fields — used by map and dashboard."""

    farm_id: str
    farm_name: Optional[str] = None
    location: Optional[str] = None
    farm_latitude: Optional[float] = None
    farm_longitude: Optional[float] = None
    total_area_acres: Optional[float] = None
    fields: list[FieldResponse] = []
    total_fields: int = 0
    total_field_area_acres: float = 0.0
    crop_distribution: dict[str, float] = Field(
        default_factory=dict,
        description="Map of crop_name → total acres planted.",
    )


# Rebuild FieldResponse so the forward reference resolves.
FieldResponse.model_rebuild()

File: Frontend/greenflora/types/field.ts
/**
 * types/field.ts
 *
 * TypeScript shapes for Fields, Crop Cycles, and Farm Summary.
 * Kept in sync with the backend's schemas/field.py.
 */

/** A crop cycle attached to a field. */
export interface CropCycle {
  id: string;
  field_id: string;
  crop_name: string;
  variety: string | null;
  crop_stage: string | null;
  planting_date: string | null;
  expected_harvest_date: string | null;
  status: "active" | "harvested" | "cancelled";
  is_demo: boolean;
}

/** A single field within a farm. */
export interface Field {
  id: string;
  farm_id: string;
  name: string;
  area_acres: number | null;
  latitude: number | null;
  longitude: number | null;
  boundary_geojson: string | null;
  soil_type: string | null;
  irrigation_method: string | null;
  status: "active" | "fallow" | "inactive";
  is_demo: boolean;
  active_crop_cycle: CropCycle | null;
}

/** Payload for creating a new field. */
export interface FieldCreate {
  name: string;
  area_acres?: number | null;
  latitude?: number | null;
  longitude?: number | null;
  boundary_geojson?: string | null;
  soil_type?: string | null;
  irrigation_method?: string | null;
  status?: "active" | "fallow" | "inactive";
}

/** Partial update payload for a field. */
export type FieldUpdate = Partial<Omit<FieldCreate, "farm_id">>;

/** Payload for creating a new crop cycle. */
export interface CropCycleCreate {
  crop_name: string;
  variety?: string | null;
  crop_stage?: string | null;
  planting_date?: string | null;
  expected_harvest_date?: string | null;
  status?: "active" | "harvested" | "cancelled";
}

/** Partial update payload for a crop cycle. */
export type CropCycleUpdate = Partial<Omit<CropCycleCreate, "field_id">>;

/** Farm summary with fields — used by dashboard + map. */
export interface FarmSummary {
  farm_id: string;
  farm_name: string | null;
  location: string | null;
  farm_latitude: number | null;
  farm_longitude: number | null;
  total_area_acres: number | null;
  fields: Field[];
  total_fields: number;
  total_field_area_acres: number;
  crop_distribution: Record<string, number>;
}
